pkg/capacityplanner/helpers.go

```diff
@@ -0,0 +1,347 @@
+	if err != nil {
+		logger.Errorf("failed to read ACR list: %s", err.Error())
+		return err
+	}
+	rh.acMap = buildACMap(acList)
+	rh.acrMap, rh.acNameToACR = buildACRMaps(acrList)
+
+	return nil
+}
+
+func (rh *ReservationHelper) removeACR(ctx context.Context, acr *acrcrd.AvailableCapacityReservation) error {
+	logger := util.AddCommonFields(ctx, rh.logger, "ReservationHelper.removeACR")
+	err := rh.client.Delete(ctx, acr)
+	if err == nil {
+		logger.Infof("ACR %s removed", acr.Name)
+		return nil
+	}
+	if k8serrors.IsNotFound(err) {
+		logger.Infof("ACR %s already removed", acr.Name)
+		return nil
+	}
+	logger.Errorf("Fail to remove ACR %s: %s", acr.Name, err.Error())
+	return err
+}
+
+func (rh *ReservationHelper) replaceACInACR(ctx context.Context, removedACR string,
+	ac, acReplacement *accrd.AvailableCapacity) error {
+	logger := util.AddCommonFields(ctx, rh.logger, "ReservationHelper.replaceACInACR")
+
+	acrToUpdate, ok := rh.acNameToACR[ac.Name]
+	if !ok {
+		logger.Infof("Can't find ACRs for AC %s", ac.Name)
+		return nil
+	}
+	for _, acrName := range acrToUpdate {
+		// ignore already removed ACR
+		if acrName == removedACR {
+			continue
+		}
+		acr, ok := rh.acrMap[acrName]
+		if !ok {
+			continue
+		}
+		replaced := false
+		for i := 0; i < len(acr.Spec.Reservations); i++ {
+			if acr.Spec.Reservations[i] == ac.Name {
+				acr.Spec.Reservations[i] = acReplacement.Name
+				replaced = true
+			}
+		}
+		if replaced {
+			err := rh.client.Update(ctx, acr)
+			if err != nil {
+				logger.Infof("Fail to update ACR %s: %s", acr.Name, err.Error())
+				return err
+			}
+			logger.Infof("ACR %s updated", acr.Name)
+		}
+	}
+	return nil
+}
+
+// NewReservationFilter returns new instance of ReservationFilter
+func NewReservationFilter(
+	acrFilter func(acr acrcrd.AvailableCapacityReservation) bool,
+	acFilter func(ac accrd.AvailableCapacity) bool) *ReservationFilter {
+	return &ReservationFilter{
+		acrFilter: acrFilter,
+		acFilter:  acFilter,
+	}
+}
+
+// ReservationFilter helper for working with ACR based reservations
+type ReservationFilter struct {
+	// additional filter for ACR
+	acrFilter func(acr acrcrd.AvailableCapacityReservation) bool
+	// additional filter for AC
+	acFilter func(ac accrd.AvailableCapacity) bool
+}
+
+// FilterByReservation returns AC which are reserved if reserved == true, or not reserved otherwise
+func (rh *ReservationFilter) FilterByReservation(reserved bool, acs []accrd.AvailableCapacity,
+	acrs []acrcrd.AvailableCapacityReservation) []accrd.AvailableCapacity {
+	acs, acrs = rh.filter(acs, acrs)
+	acInACR := buildACInACRMap(acrs)
+	return FilterACList(acs, func(ac accrd.AvailableCapacity) bool {
+		_, acIsReserved := acInACR[ac.Name]
+		if reserved {
+			// we looking for reserved ACs
+			return acIsReserved
+		}
+		// we looking for unreserved ACs
+		return !acIsReserved
+	})
+}
+
+func (rh *ReservationFilter) filter(acs []accrd.AvailableCapacity, acrs []acrcrd.AvailableCapacityReservation) (
```

Comment:
Could you please rename `rh`? It is confusing because of `ReservationHelper`.

Reply:
Done